src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,464 @@
+
+    VUM.write mciPrimary pageNr minPrimbits
+
+    lastMaxPrefix <- readSTRef mciLastMaxPrimbits
+    let clash = lastMaxPrefix == SJust minPrimbits
+    VUM.write mciClashes pageNr (Bit clash)
+    when clash $ modifySTRef' mciTieBreaker (IntMap.insert pageNr minKey)
+    writeSTRef mciLastMaxPrimbits $ SJust maxPrimbits
+  where
+      minRfbits :: Word16
+      minRfbits = topBits16 mciRangeFinderPrecision minKey
+
+      minPrimbits, maxPrimbits :: Word32
+      minPrimbits = sliceBits32 mciRangeFinderPrecision minKey
+      maxPrimbits = sliceBits32 mciRangeFinderPrecision maxKey
+
+-- | Make a full copy of the mutable compact index and return it as an immutable
+-- compact index. The 'MCompactIndex' can be safely mutated afterwards.
+freeze :: MCompactIndex s -> ST s CompactIndex
+freeze mci@MCompactIndex{..} = do
+    -- We are not guaranteed to have seen all possible range-finder bit
+    -- combinations, so we have to fill in the remainder of the rangerfinder
+    -- vector.
+    fillRangeFinderToEnd mci
+
+    ciRangeFinder <- VU.freeze mciRangeFinder
+    ciPrimary <- VU.freeze mciPrimary
+    ciClashes <- VU.freeze mciClashes
+    ciTieBreaker <- readSTRef mciTieBreaker
+    pure $ CompactIndex {
+        ciRangeFinderPrecision = mciRangeFinderPrecision
+      , ..
+      }
+
+-- | Like 'freeze', but 'unsafeFreeze' does not make full copies of underlying
+-- vectors and arrays. Do /not/ modify the 'MCompactIndex' afterwards.
+unsafeFreeze :: MCompactIndex s -> ST s CompactIndex
+unsafeFreeze mci@MCompactIndex{..} = do
+    -- We are not guaranteed to have seen all possible range-finder bit
+    -- combinations, so we have to fill in the remainder of the rangerfinder
+    -- vector.
+    fillRangeFinderToEnd mci
+
+    ciRangeFinder <- VU.unsafeFreeze mciRangeFinder
+    ciPrimary <- VU.unsafeFreeze mciPrimary
+    ciClashes <- VU.unsafeFreeze mciClashes
+    ciTieBreaker <- readSTRef mciTieBreaker
+    pure $ CompactIndex {
+        ciRangeFinderPrecision = mciRangeFinderPrecision
+      , ..
+      }
+
+-- | Fill the remainder of the range-finder vector.
+fillRangeFinderToEnd :: MCompactIndex s -> ST s ()
+fillRangeFinderToEnd MCompactIndex{..} = do
+    lastMinRfbits <- readSTRef mciLastMinRfbits
+    let lb = smaybe NoBound (BoundExclusive . fromIntegral) lastMinRfbits
+        ub = NoBound
+        x  = fromIntegral $ VUM.length mciPrimary
+    writeRange mciRangeFinder lb ub x
+    writeSTRef mciLastMinRfbits (SJust $ 2 ^ mciRangeFinderPrecision)
+
+{-------------------------------------------------------------------------------
+  Chunks
+-------------------------------------------------------------------------------}
+
+data Chunk = Chunk {
+    cPrimary :: !(VU.Vector Word32)
+  , cClashes :: !(VU.Vector Bit)
+  }
+
+data FinalChunk = FinalChunk {
+    fcRangeFinder          :: !(VU.Vector Word32)
+  , fcRangeFinderPrecision :: !Int
+  , fcTieBreaker           :: !(IntMap SerialisedKey)
+  }
```

Comment:
I have no idea how this is supposed to work or be used. How do we make a chunk? How do these chunks relate to the use of `new`? I don't see how we do incremental construction, i.e. without knowing in advance the number of pages in the index.